File: src/main/kotlin/io/energyconsumptionoptimizer/mapservice/application/FloorPlanServiceImpl.kt
package io.energyconsumptionoptimizer.mapservice.application

import io.energyconsumptionoptimizer.mapservice.domain.FloorPlan
import io.energyconsumptionoptimizer.mapservice.domain.errors.FlorPlanFormatNotValidException
import io.energyconsumptionoptimizer.mapservice.domain.ports.FloorPlanService
import io.energyconsumptionoptimizer.mapservice.domain.ports.HouseMapRepository
import org.xml.sax.InputSource
import java.io.StringReader
import javax.xml.parsers.DocumentBuilderFactory

class FloorPlanServiceImpl(
    private val repository: HouseMapRepository,
) : FloorPlanService {
    override suspend fun createFloorPlan(floorPlanSVG: String): FloorPlan {
        if (!isValidSVG(floorPlanSVG)) {
            throw FlorPlanFormatNotValidException()
        }

        val floorPlan = FloorPlan(floorPlanSVG)
        return repository.saveFloorPlan(floorPlan)
    }

    override suspend fun getFloorPlan(): FloorPlan? = repository.findFloorPlan()

    private fun isValidSVG(svg: String): Boolean =
        try {
            val factory = DocumentBuilderFactory.newInstance()

            val builder = factory.newDocumentBuilder()
            val inputSource = InputSource(StringReader(svg))
            val document = builder.parse(inputSource)

            val root = document.documentElement
            root != null && root.tagName.equals("svg", ignoreCase = true)
        } catch (_: Exception) {
            false
        }
}

File: src/application/FloorPlanServiceImpl.ts
import type { HouseMapRepository } from "@domain/ports/HouseMapRepository";
import type { FloorPlanService } from "@application/inbound/FloorPlanService";
import { FloorPlan } from "@domain/values/FloorPlan";
import { InvalidFloorPlanError } from "@domain/errors";
import { BusinessMetrics } from "@application/outbound/BusinessMetrics";

const SVG_ROOT_TAG = /<svg[\s>]/i;

export class FloorPlanServiceImpl implements FloorPlanService {
  readonly #repository: HouseMapRepository;
  readonly #metrics: BusinessMetrics;

  constructor(repository: HouseMapRepository, metrics: BusinessMetrics) {
    this.#repository = repository;
    this.#metrics = metrics;
  }

  async createFloorPlan(floorPlanSVG: string): Promise<FloorPlan | Error> {
    if (!this.#isValidSVG(floorPlanSVG)) {
      return new InvalidFloorPlanError();
    }

    const floorPlan = FloorPlan.from(floorPlanSVG);
    if (floorPlan instanceof Error) return floorPlan;

    await this.#repository.saveFloorPlan(floorPlan);

    this.#metrics.recordFloorPlanCreation();

    return floorPlan;
  }

  async getFloorPlan(): Promise<FloorPlan | null> {
    return this.#repository.findFloorPlan();
  }

  /**
   * Validates that the SVG string has a root <svg> element.
   * We use a lightweight regex check to avoid pulling in an XML parser.
   */
  #isValidSVG(svg: string): boolean {
    return SVG_ROOT_TAG.test(svg.trimStart());
  }
}
